transport/http/agent.go

```diff
@@ -0,0 +1,123 @@
+package http
+
+import (
+	"net/http"
+	"sync"
+
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/peer/hostport"
+)
+
+// Agent keeps track of http peers and the associated client with which the peer will call into.
+type Agent struct {
+	sync.Mutex
+
+	client    *http.Client
+	peerNodes map[string]*peerNode
+}
+
+// peerNode keeps track of a HostPortPeer and any subscribers referencing it
+type peerNode struct {
+	peer       *hostport.Peer
+	references map[transport.PeerSubscriber]bool
+}
+
+// NewDefaultAgent creates an http agent with the default parameters
+func NewDefaultAgent() *Agent {
+	return NewAgent(&defaultConfig)
+}
+
+// NewAgent creates a new http agent for managing peers and sending requests
+func NewAgent(cfg *outboundConfig) *Agent {
+	return &Agent{
+		client:    buildClient(cfg),
+		peerNodes: make(map[string]*peerNode),
+	}
+}
+
+// GetClient gets the http client that should be used for making requests
+func (a *Agent) GetClient() *http.Client {
+	return a.client
+}
+
+// RetainPeer gets or creates a Peer for the specificed PeerList
+func (a *Agent) RetainPeer(pid transport.PeerIdentifier, sub transport.PeerSubscriber) (transport.Peer, error) {
+	a.Mutex.Lock()
+	defer a.Mutex.Unlock()
+
+	hppid, ok := pid.(hostport.PeerIdentifier)
+	if !ok {
+		return nil, errors.ErrInvalidPeerType{
+			ExpectedType:   "hostport.PeerIdentifier",
+			PeerIdentifier: pid,
+		}
+	}
+
+	node := a.getOrCreatePeerNode(hppid)
+
+	node.references[sub] = true
+
+	return node.peer, nil
+}
+
+func (a *Agent) getOrCreatePeerNode(pid hostport.PeerIdentifier) *peerNode {
+	if node, ok := a.peerNodes[pid.Identifier()]; ok {
```

Comment:
actually, after further investigation, this can lead to issues with casting, if we pass in a *hostport.Peer as a transport.PeerIdentifier and try to convert it to a hostport.PeerIdentifier we will get a casting error, it's easier to just use the Identifier() func for now